neqo-http3/src/recv_message.rs

```diff
@@ -132,32 +135,47 @@ impl RecvMessage {
         Ok(())
     }
 
-    fn add_headers(&mut self, headers: Option<Vec<Header>>, fin: bool, decoder: &mut QPackDecoder) {
-        if fin {
-            self.conn_events.header_ready(self.stream_id, headers, true);
-            self.set_closed(decoder);
-        } else {
-            self.conn_events
-                .header_ready(self.stream_id, headers, false);
-            self.state = RecvMessageState::WaitingForData {
-                frame_reader: HFrameReader::new(),
-            };
+    fn add_headers(
+        &mut self,
+        headers: Vec<Header>,
+        fin: bool,
+        decoder: &mut QPackDecoder,
+    ) -> Res<()> {
+        match (fin, self.is_interim(&headers)?) {
```

Comment:
I think that your match here isn't really helping much.  Every arm of this is essentially the same code.  I think that you can do:

```rust
let interim = Self::is_interim(&headers)?;
if fin && interim {
  return Err(Error:HttpGeneralProtocolStream);
}
self.conn_events.header_ready(self.stream_id, headers, true, fin);
self.state = if interim {
RecvMessageState::WaitingForResponseHeaders {
                    frame_reader: HFrameReader::new(),
                }
} else {
 RecvMessageState::WaitingForData {
                    frame_reader: HFrameReader::new(),
                }
};
if fin {
  self.set_closed(decoder);
}